File: Wells/Source/VievModels/Graphics/PiperSchoellerGraphicViewModel.vb
Imports Wells.ViewBase
Imports Wells.YPFModel

Public Class PiperSchoellerGraphicViewModel
    Inherits BaseViewModel

    Private Const CalciumMeq As Double = 20
    Private Const PotassiumMeq As Double = 39
    Private Const SodiumMeq As Double = 23
    Private Const MagnesiumMeq As Double = 12.15
    Private Const CarbonateMeq As Double = 61
    Private Const SulfatesMeq As Double = 48
    Private Const ChloridesMeq As Double = 35.5

    Dim _Random As New Random()

    Private _Dialog As IPiperSchoellerGraphicView
    Private _PiperSchollerPoints As List(Of PiperSchollerData)
    Private _SelectedPoint As PiperSchollerData

    ReadOnly Property PiperSchollerPoints As List(Of PiperSchollerData)
        Get
            Return _PiperSchollerPoints
        End Get
    End Property

    Property SelectedPoint As PiperSchollerData
        Get
            Return _SelectedPoint
        End Get
        Set
            SetValue(_SelectedPoint, Value)
        End Set
    End Property

    Sub New(wells As IEnumerable(Of Well))
        MyBase.New(Nothing)
        Initialize()
        InitializeData(wells)
    End Sub

    Sub New(analises As IEnumerable(Of WaterAnalysis))
        MyBase.New(Nothing)
        Initialize()
        InitializeData(analises)
    End Sub

    Private Sub InitializeData(analises As IEnumerable(Of WaterAnalysis))
        _PiperSchollerPoints = New List(Of PiperSchollerData)

        Dim group = analises.GroupBy(Function(a) a.Well)

        For Each g In group
            For Each a In g
                Dim aColor = GetRandomColor()
                _PiperSchollerPoints.Add(CalculatePercentage(a, aColor))
            Next
        Next

    End Sub

    Private Sub InitializeData(wells As IEnumerable(Of Well))
        _PiperSchollerPoints = New List(Of PiperSchollerData)

        For Each w In wells
            Dim aColor = GetRandomColor()
            For Each a In w.WaterAnalyses
                _PiperSchollerPoints.Add(CalculatePercentage(a, aColor))
            Next
        Next

    End Sub

    Private Function GetRandomColor() As Color
        Dim names = [Enum].GetValues(GetType(System.Drawing.KnownColor))
        Dim randomColorName = names(_Random.Next(names.Length))
        Dim randomColor = System.Drawing.Color.FromKnownColor(randomColorName)
        Return Color.FromArgb(randomColor.A, randomColor.R, randomColor.G, randomColor.B)
    End Function

    Private Function CalculatePercentage(analisys As WaterAnalysis, color As Color) As PiperSchollerData
        Dim cal = analisys.Calcium / CalciumMeq
        Dim pot = analisys.Potassium / PotassiumMeq
        Dim sod = analisys.Sodium / SodiumMeq
        Dim mag = analisys.Magnesium / MagnesiumMeq
        Dim totalCations = cal + pot + sod + mag

        Dim carb = analisys.CarbonateAlkalinity / CarbonateMeq
        Dim sulf = analisys.Sulfates / SulfatesMeq
        Dim chlo = analisys.Chlorides / ChloridesMeq
        Dim totalAnions = carb + sulf + chlo

        Dim psData As New PiperSchollerData() With {
        .Calcium = cal / totalCations * 100,
        .Magnesium = mag / totalCations * 100,
        .PotassiumAndSodium = (sod + pot) / totalCations * 100,
        .Carbonate = carb / totalAnions * 100,
        .Chlorides = chlo / totalAnions * 100,
        .Sulfates = sulf / totalAnions * 100,
        .PointColor = color,
        .Label = $"{analisys.WellName} ({analisys.ToString()})",
        .IsVisible = True
        }

        Return psData
    End Function

    Protected Overrides Sub OnSetView(view As IView)
        MyBase.OnSetView(view)
        _Dialog = CType(view, IPiperSchoellerGraphicView)
    End Sub

    Protected Overrides Sub SetValidators()

    End Sub

    Protected Overrides Sub SetCommandUpdates()
        AddCommands(NameOf(SelectedPoint), {ChangeColorCommand})
    End Sub

    ReadOnly Property ChangeColorCommand As ICommand = New RelayCommand(Sub()
                                                                            Dim currentColor = System.Drawing.Color.FromArgb(SelectedPoint.PointColor.A,
                                                                                                                             SelectedPoint.PointColor.R,
                                                                                                                             SelectedPoint.PointColor.G,
                                                                                                                             SelectedPoint.PointColor.B)
                                                                            SelectedPoint.PointColor = _Dialog.ShowColorDialog(currentColor)
                                                                            _Dialog.CreateGraphics()
                                                                        End Sub, Function() SelectedPoint IsNot Nothing, AddressOf OnError)

    ReadOnly Property SaveImageCommand As ICommand = New RelayCommand(Sub()
                                                                          _Dialog.SaveImage()
                                                                      End Sub, Function() True, AddressOf OnError)
End Class

File: Wells/src/ViewModels/Graphics/PiperSchoellerGraphicViewModel.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Windows.Input;
using System.Windows.Media;
using Wells.BaseView;
using Wells.BaseView.ViewInterfaces;
using Wells.BaseView.ViewModel;
using Wells.View.Graphics;
using Wells.Model;

namespace Wells.View.ViewModels
{
    public class PiperSchoellerGraphicViewModel : BaseViewModel
    {
        private const double CalciumMeq = 20.0;
        private const double PotassiumMeq = 39.0;
        private const double SodiumMeq = 23.0;
        private const double MagnesiumMeq = 12.15;
        private const double BicarbonateMeq = 61.0;
        private const double SulfatesMeq = 48.0;
        private const double ChloridesMeq = 35.5;

        private readonly Random _Random = new Random();

        IPiperSchoellerGraphicView _Dialog;

        private bool showZones;
        private int pointSize;
        private PiperSchollerData selectedPoint;

        public List<PiperSchollerData> PiperSchollerPoints { get; private set; }

        public PiperSchollerData SelectedPoint { get => selectedPoint; set { SetValue(ref selectedPoint, value); } }

        public bool ShowZones { get => showZones; set { SetValue(ref showZones, value); _Dialog?.CreateGraphics(); } }


        public int PointSize
        {
            get => pointSize;
            set
            {
                if (value >= 10 && value <= 1000)
                {
                    SetValue(ref pointSize, value);
                    _Dialog?.CreateGraphics();
                }
            }
        }

        public PiperSchoellerGraphicViewModel(IEnumerable<Well> wells) : base(null)
        {
            showZones = true;
            pointSize = 100;
            Initialize();
            InitializeData(wells);
        }

        public PiperSchoellerGraphicViewModel(IEnumerable<WaterAnalysis> analyses) : base(null)
        {
            showZones = true;
            pointSize = 100;
            Initialize();
            InitializeData(analyses);
        }


        void InitializeData(IEnumerable<WaterAnalysis> analyses)
        {
            PiperSchollerPoints = new List<PiperSchollerData>();

            var group = analyses.GroupBy(a => a.Well);

            foreach (var g in group)
            {
                foreach (var a in g)
                {
                    var aColor = SharedBaseView.GetRandomColor(_Random);
                    PiperSchollerPoints.Add(CalculatePercentage(a, aColor));
                }
            }

        }

        void InitializeData(IEnumerable<Well> wells)
        {
            PiperSchollerPoints = new List<PiperSchollerData>();

            foreach (var w in wells)
            {
                var aColor = SharedBaseView.GetRandomColor(_Random);
                foreach (var a in w.WaterAnalyses)
                {
                    PiperSchollerPoints.Add(CalculatePercentage(a, aColor));
                }
            }

        }

        protected override void SetCommandUpdates()
        {
            Add(nameof(SelectedPoint), ChangeColorCommand);
        }

        PiperSchollerData CalculatePercentage(WaterAnalysis analysis, Color color)
        {
            var cal = analysis.Calcium / CalciumMeq;
            var pot = analysis.Potassium / PotassiumMeq;
            var sod = analysis.Sodium / SodiumMeq;
            var mag = analysis.Magnesium / MagnesiumMeq;
            var totalCations = cal + pot + sod + mag;

            var bicarb = analysis.BicarbonateAlkalinity / BicarbonateMeq;
            var sulf = analysis.Sulfates / SulfatesMeq;
            var chlo = analysis.Chlorides / ChloridesMeq;
            var totalAnions = bicarb + sulf + chlo;

            var psData = new PiperSchollerData
            {
                Calcium = cal / totalCations * 100,
                Magnesium = mag / totalCations * 100,
                PotassiumAndSodium = (sod + pot) / totalCations * 100,
                Carbonates = bicarb / totalAnions * 100,
                Chlorides = chlo / totalAnions * 100,
                Sulfates = sulf / totalAnions * 100,
                PointColor = color,
                Label = $"{analysis.WellName} ({analysis.ToString()})",
                IsVisible = true
            };

            return psData;
        }

        protected override void OnSetView(IView view)
        {
            base.OnSetView(view);
            _Dialog = (view as IPiperSchoellerGraphicView);
        }



        public ICommand SaveImageCommand
        {
            get
            {
                return new RelayCommand((param) =>
                {
                    _Dialog.SaveImage("Piper");
                }, (obj) => true, OnError);
            }
        }

        public ICommand ChangeColorCommand
        {
            get
            {
                return new RelayCommand((param) =>
                {
                    var currentColor = SelectedPoint.PointColor.ToDrawingColor();
                    SelectedPoint.PointColor = SharedBaseView.ShowColorDialog(currentColor);
                    _Dialog.CreateGraphics();
                }, (obj) => SelectedPoint != null, OnError);
            }
        }

    }
}
